src/pgduckdb_node.cpp

```diff
@@ -203,60 +203,69 @@ ExecuteQuery(DuckdbScanState *state) {
+		bool already_executed = duckdb_scan_state->is_executed;
+		if (!already_executed) {
+			ExecuteQuery(duckdb_scan_state);
+		}
 
-	/* MemoryContext used for allocation */
-	old_context = MemoryContextSwitchTo(duckdb_scan_state->css.ss.ps.ps_ExprContext->ecxt_per_tuple_memory);
+		if (duckdb_scan_state->fetch_next) {
+			duckdb_scan_state->current_data_chunk = duckdb_scan_state->query_results->Fetch();
+			duckdb_scan_state->current_row = 0;
+			duckdb_scan_state->fetch_next = false;
+			if (!duckdb_scan_state->current_data_chunk || duckdb_scan_state->current_data_chunk->size() == 0) {
+				MemoryContextReset(duckdb_scan_state->css.ss.ps.ps_ExprContext->ecxt_per_tuple_memory);
+				ExecClearTuple(slot);
+				return slot;
+			}
+		}
 
-	for (idx_t col = 0; col < duckdb_scan_state->column_count; col++) {
-		// FIXME: we should not use the Value API here, it's complicating the LIST conversion logic
-		auto value = duckdb_scan_state->current_data_chunk->GetValue(col, duckdb_scan_state->current_row);
-		if (value.IsNull()) {
-			slot->tts_isnull[col] = true;
-		} else {
-			slot->tts_isnull[col] = false;
-			if (!pgduckdb::ConvertDuckToPostgresValue(slot, value, col)) {
-				CleanupDuckdbScanState(duckdb_scan_state);
-				throw duckdb::ConversionException("Value conversion failed");
+		MemoryContextReset(duckdb_scan_state->css.ss.ps.ps_ExprContext->ecxt_per_tuple_memory);
+		ExecClearTuple(slot);
+
+		/* MemoryContext used for allocation */
+		old_context = MemoryContextSwitchTo(duckdb_scan_state->css.ss.ps.ps_ExprContext->ecxt_per_tuple_memory);
+
+		for (idx_t col = 0; col < duckdb_scan_state->column_count; col++) {
+			// FIXME: we should not use the Value API here, it's complicating the LIST conversion logic
+			auto value = duckdb_scan_state->current_data_chunk->GetValue(col, duckdb_scan_state->current_row);
+			if (value.IsNull()) {
+				slot->tts_isnull[col] = true;
+			} else {
+				slot->tts_isnull[col] = false;
+				if (!pgduckdb::ConvertDuckToPostgresValue(slot, value, col)) {
+					throw duckdb::ConversionException("Value conversion failed");
+				}
 			}
 		}
-	}
 
-	MemoryContextSwitchTo(old_context);
+		MemoryContextSwitchTo(old_context);
 
-	duckdb_scan_state->current_row++;
-	if (duckdb_scan_state->current_row >= duckdb_scan_state->current_data_chunk->size()) {
-		duckdb_scan_state->current_data_chunk.reset();
-		duckdb_scan_state->fetch_next = true;
-	}
+		duckdb_scan_state->current_row++;
+		if (duckdb_scan_state->current_row >= duckdb_scan_state->current_data_chunk->size()) {
+			duckdb_scan_state->current_data_chunk.reset();
+			duckdb_scan_state->fetch_next = true;
+		}
 
-	ExecStoreVirtualTuple(slot);
-	return slot;
+		ExecStoreVirtualTuple(slot);
+		return slot;
+	} catch (std::exception &ex) {
+		/*
+		 * In case any error happens we need to still cleanup the scan state,
+		 * otherwise we do not clean up the prepared statement and various
+		 * other DuckDB objects.
+		 */
+		CleanupDuckdbScanState(duckdb_scan_state);
```

Comment:
The usual way to do that is to use RAII. We could add this trivial struct above in the file:

```c++
struct ScopedDuckDBScanState {
  ScopedDuckDBScanState(_state) : state(_state);
  ~ScopedDuckDBScanState() {
     CleanupDuckdbScanState(state);
  }

private:
  DuckdbScanState *state;
}
```

And then at the beginning of this function
```c++
  DuckdbScanState *duckdb_scan_state = (DuckdbScanState *)node;
  ScopedDuckDBScanState(duckdb_scan_state);
```

Reply:
But that would also clean it up on success right? We don't want to do that.

Reply:
Updated the comment to reflect this, because indeed it's not immediately obvious.